### ISSUE TITLE
Expand breadcrumbs to show other elements (other than file symbols)

### ISSUE THREAD

**Issue by alefragnani** [Microsoft/vscode#65515]
Hi,

Would it be possible to expand breadcrumbs to support other elements, like `decorators`, `diagnostics`, or any other _provided_ elements (via an API similar to TreeView)? 

The idea is adding [Bookmarks](
https://marketplace.visualstudio.com/items?itemName=alefragnani.Bookmarks) in there.

Thank you

**Comment by jrieken** [comment]
Not sure I understand. Do you want to add data onto existing breadcrumb items or provide your own items? The latter is possible by implementing a `DocumentSymbolProvider`

**Comment by alefragnani** [comment]
Hi @jrieken ,

I would like to provide my own items, which today I use `Decorators`, not `Symbols` or any other `Provider` interface. 

I use `Decorators` because that’s what Bookmarks do _visually_. It provides visual feedback in `Gutter` (icon), `OverviewRulerLane` and `Line`, to _show_ where the bookmarks are located in the document.  The bookmarks itself are attached to the _line number_, not the source code itself, like `Diagnostics`, but are updated (moved up or down) when changes are made to the source code.

Would I replace `Decorator ` with other _compatible_ Interface, in order to be available in Breadcrumbs, or would Breadcrumbs be updated to support `Decorators`/other elements?

Thanks for your help

**Comment by jrieken** [comment]
You mean `setDecorations` from the `TextEditor`, right? That something purely visual and not available outside of an editor. Tho, the document symbols provider api says Symbol but you can use it for whatever in theory. Tho, it will show up in the outline as well and the whole UX is geared towards working with symbols.

**Comment by alefragnani** [comment]
Exactly! I thought about `Symbol` but none of the `SymbolKind` appeared to fit. Maybe if you introduce a `Custom` Kind, which I could also provide its `Icon`, would work better. 

I also thought about `Diagnostics` and `Definition` but both also has their limitations. And I wonder what could happen with the current UI when  _real definitions_ are merged with bookmarks 😕 

Thank you

**Comment by jrieken** [comment]
> with the current UI when real definitions are merged with bookmarks

We don't really merge them but use the extension name as 'root' node. For instance, this is how it looks for me with TypeScript and Unit Test Navigate extension.


<img width="499" alt="screenshot 2018-12-28 at 09 12 19" src="https://user-images.githubusercontent.com/1794099/50508105-b2996600-0a80-11e9-8874-c55e69e5a8b9.png">

**Comment by jrieken** [comment]
Note that when registering a provider you can provide a name which is used in breadcrumbs and outline (the default would be your extension name)
